lib/icu.js

```diff
@@ -0,0 +1,199 @@
+'use strict';
+
+if (!process.binding('config').hasIntl)
+  throw new Error('This Node.js binary was built without ICU support.');
+
+const Buffer = require('buffer').Buffer;
+const normalizeEncoding = require('internal/util').normalizeEncoding;
+const icu = process.binding('icu');
+
+// Maps the supported transcoding conversions. The top key is the from_enc,
+// the child key is the to_enc. The value is the transcoding function to.
+const conversions = {
+  'ascii': {
+    'latin1': (source) => {
+      return Buffer.from(source);
+    },
+    'utf8': (source) => {
+      return Buffer.from(source);
+    },
+    'utf16le': (source) => {
+      return icu.convertToUcs2('us-ascii', source);
+    }
+  },
+  'latin1': {
+    'ascii': (source) => {
+      return icu.convert('us-ascii', 'iso8859-1', source);
+    },
+    'utf8': (source) => {
+      return icu.convert('utf-8', 'iso8859-1', source);
+    },
+    'utf16le': (source) => {
+      return icu.convertToUcs2('iso8859-1', source);
+    }
+  },
+  'utf8': {
+    'ascii': (source) => {
+      return icu.convert('us-ascii', 'utf-8', source);
+    },
+    'latin1': (source) => {
+      return icu.convert('iso-8859-1', 'utf-8', source);
+    },
+    'utf16le': icu.convertToUcs2FromUtf8,
+  },
+  'utf16le': {
+    'ascii': (source) => {
+      if (source.length % 2 !== 0)
+        throw new TypeError('Invalid UCS2 Buffer');
+      return icu.convertFromUcs2('us-ascii', source);
+    },
+    'latin1': (source) => {
+      if (source.length % 2 !== 0)
+        throw new TypeError('Invalid UCS2 Buffer');
+      return icu.convertFromUcs2('iso-8859-1', source);
+    },
+    'utf8': (source) => {
+      if (source.length % 2 !== 0)
+        throw new TypeError('Invalid UCS2 Buffer');
+      return icu.convertToUtf8FromUcs2(source);
+    }
+  }
+};
+
+// Transcodes the Buffer from one encoding to another, returning a new
+// Buffer instance.
+exports.transcode = function transcode(source, from_enc, to_enc) {
+  if (!source || !(source.buffer instanceof ArrayBuffer))
+    throw new TypeError('"source" argument must be a Buffer');
+  if (source.length === 0) return Buffer.alloc(0);
+
+  from_enc = normalizeEncoding(from_enc);
+  to_enc = normalizeEncoding(to_enc);
+
+  if (from_enc === to_enc)
+    return Buffer.from(source);
+
+  const cnv_from = conversions[from_enc];
+
+  if (cnv_from) {
+    const cnv_to = cnv_from[to_enc];
+    if (cnv_to)
+      return cnv_to(source);
+  }
+  throw new TypeError(`Unsupported conversion: ${from_enc} to ${to_enc}`);
+};
+
+// Returns the unicode codepoint at a specific offset location in the buffer,
+// based on the given encoding
+exports.codePointAt = function(buf, pos, encoding) {
+  if (!buf || !(buf.buffer instanceof ArrayBuffer))
+    throw new TypeError('First argument must be a Buffer');
```

Comment:
this necessary as it's the first check in C++, and the fact that this can be fooled fairly easily?